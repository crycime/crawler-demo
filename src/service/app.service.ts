import { Injectable } from '@nestjs/common';
import * as playwright from 'playwright-aws-lambda';
import * as _ from 'lodash';
import { AppCrawlerGetResDto } from '../dto/appCrawlerGetRes.dto';

@Injectable()
export class AppService {
  async crawler(): Promise<AppCrawlerGetResDto> {
    let browser = null;
    try {
      browser = await playwright.launchChromium({ headless: true });
      const context = await browser.newContext();
      const page = await context.newPage();
      await page.goto(
        'https://www.latenitex.com/products/faux-suede-hooded-puffer-jacket?variant=43895596613871',
      );
      const liElements = await page.$$(
        '#Slider-Thumbnails-template--17140330889455__main li',
      );
      const imageUrlList: any[] = [];
      await Promise.all(
        _.map(liElements, async (liElement) => {
          // 获取每个li下的img元素的src属性
          const imgSrc = await liElement.$eval('img', (img) => img.src);
          imageUrlList.push(imgSrc);
        }),
      );
      //
      const title = await page
        .locator(
          '#ProductInfo-template--17140330889455__main > div.product__title > h1',
        )
        .innerText();
      //
      const originalPrice = await page
        .locator(
          '#price-template--17140330889455__main > div > div > div.price__sale > span:nth-child(2) > s',
        )
        .innerText();
      //
      const preferentialPrice = await page
        .locator(
          '#price-template--17140330889455__main > div > div > div.price__sale > span.price-item.price-item--sale.price-item--last',
        )
        .innerText();
      //
      const colorList: any[] = [];
      const colorElements = await page.$$(
        '#variant-radios-template--17140330889455__main > fieldset:nth-child(1) label',
      );
      await Promise.all(
        colorElements.map(async (labelElement) => {
          const labelText = await labelElement.innerText();
          colorList.push(labelText);
        }),
      );
      //
      const sizeList: any[] = [];
      const sizeElements = await page.$$(
        '#variant-radios-template--17140330889455__main > fieldset:nth-child(2) label',
      );
      await Promise.all(
        sizeElements.map(async (labelElement) => {
          const labelText = await labelElement.innerText();
          sizeList.push(labelText);
        }),
      );
      //
      const descriptionList: any[] = [];
      const descriptionElements = await page.$$(
        '#ProductInfo-template--17140330889455__main > div.product__description.rte.quick-add-hidden > p',
      );
      await Promise.all(
        descriptionElements.map(async (labelElement) => {
          const labelText = await labelElement.innerText();
          descriptionList.push(labelText);
        }),
      );
      //
      await page.waitForSelector(
        '#ProductInfo-template--17140330889455__main > div.product__description.rte.quick-add-hidden > div.ks-chart-container.sizing-chart-container > div > div > div.ks-table-content-wrapper.ks-toggle-style-1 > div.ks-table-wrapper > table > tbody > tr:nth-child(1)',
        { state: 'visible' },
      );
      const rows = await page.$$(
        '#ProductInfo-template--17140330889455__main > div.product__description.rte.quick-add-hidden > div.ks-chart-container.sizing-chart-container > div > div > div.ks-table-content-wrapper.ks-toggle-style-1 > div.ks-table-wrapper > table > tbody > .ks-table-row',
      );
      const data = await Promise.all(
        rows.map(async (row) => {
          const columns = await row.$$('.ks-table-cell');
          return await Promise.all([
            page.evaluate((el) => el.textContent.trim(), columns[0]),
            page.evaluate((el) => el.textContent.trim(), columns[1]),
            page.evaluate((el) => el.textContent.trim(), columns[2]),
            page.evaluate((el) => el.textContent.trim(), columns[3]),
          ]);
        }),
      );
      const keys = data[0];
      const measure: any = _.map(data.slice(1), (row) =>
        _.zipObject(keys, row),
      );
      return {
        imageUrlList,
        title,
        originalPrice,
        preferentialPrice,
        colorList,
        sizeList,
        measure,
        descriptionList,
      };
    } catch (e: any) {
      return { err: e.message, stack: e.stack } as any;
    } finally {
      if (browser) {
        await browser.close();
      }
    }
  }
}
