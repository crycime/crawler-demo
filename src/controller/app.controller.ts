import { Controller, Get } from '@nestjs/common';
import { AppService } from '../service/app.service';
import { AppCrawlerGetResDto } from '../dto/appCrawlerGetRes.dto';

@Controller('/app')
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get('/crawler')
  crawler(): Promise<AppCrawlerGetResDto> {
    return this.appService.crawler();
  }
}
