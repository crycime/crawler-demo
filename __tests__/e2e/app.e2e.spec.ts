import { HttpStatus, INestApplication } from '@nestjs/common';
import request from 'supertest';
import { Test, TestingModule } from '@nestjs/testing';
import { AppController } from '../../src/controller/app.controller';
import { AppService } from '../../src/service/app.service';

const path = '/app/crawler';

describe(`GET ${path}`, () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleBuilder = Test.createTestingModule({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    });
    const compiledApp: TestingModule = await moduleBuilder.compile();
    const appModule = compiledApp.createNestApplication();
    app = await appModule.init();
  });

  describe('success cases', () => {
    it('should return crawler response', async () => {
      const response = await request(app.getHttpServer()).get(path);
      expect(response.status).toEqual(HttpStatus.OK);
      expect(response.body).toMatchSnapshot();
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await app.close();
  });
});
