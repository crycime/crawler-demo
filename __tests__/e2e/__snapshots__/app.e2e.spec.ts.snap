// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /app/crawler success cases should return crawler response 1`] = `
Object {
  "colorList": Array [
    "grey",
    "black",
    "brown",
  ],
  "descriptionList": Array [
    "Stay warm in style with our Faux Suede Hooded Puffer Jacket. It features a unique faux suede material, puffer design, and hooded and zipper closure for extra comfort",
    "Â Model is 5'9 wearing 2XL",
  ],
  "imageUrlList": Array [
    "https://www.latenitex.com/cdn/shop/products/1709514.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/1709509.jpg?v=1681225942&width=1946",
    "https://www.latenitex.com/cdn/shop/products/O1CN011B35HS25oUC611F92__2208113727573_1.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/1709512.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/1709511.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/O1CN01sDMe6T25oUC5rMXZX__2208113727573.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/O1CN01BtSUou25oUBxlvma4__2208113727573.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/1709513.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/products/1709503.jpg?v=1682065442&width=1946",
    "https://www.latenitex.com/cdn/shop/files/O1CN01TfZZqM25oUByNOdaq__2208113727573_480x480_cd9bdcec-c43b-4b4a-abaf-74869511166d.jpg?v=1690109087&width=1946",
    "https://www.latenitex.com/cdn/shop/files/O1CN01kB6DCi25oUC2lR1sr__2208113727573_480x480_eaf325b7-8276-4aba-9038-b351299d29ad.jpg?v=1690109100&width=1946",
  ],
  "measure": Array [
    Object {
      "CHEST": "48.5",
      "LENGTH": "27.5",
      "SIZE": "S",
      "SLEEVE": "30",
    },
    Object {
      "CHEST": "50.5",
      "LENGTH": "28.5",
      "SIZE": "M",
      "SLEEVE": "30.5",
    },
    Object {
      "CHEST": "52",
      "LENGTH": "29",
      "SIZE": "L",
      "SLEEVE": "31.5",
    },
    Object {
      "CHEST": "53.5",
      "LENGTH": "30",
      "SIZE": "XL",
      "SLEEVE": "32",
    },
    Object {
      "CHEST": "55",
      "LENGTH": "30.5",
      "SIZE": "2XL",
      "SLEEVE": "33",
    },
    Object {
      "CHEST": "56.5",
      "LENGTH": "31.5",
      "SIZE": "3XL",
      "SLEEVE": "34",
    },
  ],
  "originalPrice": "$199.00",
  "preferentialPrice": "$139.30",
  "sizeList": Array [
    "S",
    "M",
    "L",
    "XL",
    "2XL",
    "3XL",
  ],
  "title": "Faux Suede Hooded Puffer Jacket",
}
`;
